src/coordinator/api/v1/handler/prometheus/native/read.go

```diff
@@ -128,20 +131,104 @@ func (h *PromReadHandler) parseRequest(r *http.Request) (string, *handler.ParseE
 	return targetQueries[0], nil
 }
 
-func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, req string, params *prometheus.RequestParams) ([]ts.Series, error) {
-	ctx, cancel := context.WithTimeout(reqCtx, params.Timeout)
+func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, req string, timeout time.Duration) ([]ts.Series, error) {
+	ctx, cancel := context.WithTimeout(reqCtx, timeout)
 	defer cancel()
 
-	opts := &executor.EngineOptions{}
+	opts := &executor.EngineOptions{
+		Now: time.Now(),
+	}
 	// Detect clients closing connections
 	abortCh, _ := handler.CloseWatcher(ctx, w)
 	opts.AbortCh = abortCh
 
-	_, err := promql.Parse(req)
+	parser, err := promql.Parse(req)
 	if err != nil {
 		return nil, err
 	}
 
-	// todo(braskin): implement query execution
-	return nil, errors.New("not implemented")
+	// Results is closed by execute
+	results := make(chan executor.Query)
+	// Block series slices are sorted by start time
+	seriesMap := make(map[string][]block.Series)
+	go h.engine.ExecuteExpr(ctx, parser, opts, results)
+
+	for result := range results {
+		if result.Err != nil {
+			return nil, err
+		}
+
+		blocks := result.Result.Blocks()
+		// TODO: Stream blocks to client
+		for blk := range blocks {
+			iter := blk.SeriesIter()
+			for iter.Next() {
+				insertSeriesInMap(iter.Current(), seriesMap)
+			}
+
+			blk.Close()
```

Comment:
Don't close here; it will free all the underlying bytes in the block causing empty data. Caller will have to close it after marshalling to json

Reply:
hmm need to figure out what to do here. The caller doesn't have reference to block. Maybe this function can marshal to json.